install_sct

```diff
@@ -213,13 +213,13 @@ function check_requirements() {
 function get_shell_rc_path() {
   if [[ "$SHELL" == *"bash"* ]]; then
     THE_RC="bash"
-    RC_FILE_PATH="$HOME/.bashrc"
+    RC_FILE_PATH="$HOME/.bash_profile"
```

Comment:
ah, we need to be careful here (i think). E.g., if someone reinstalls SCT, the older SCT path will still be declared in the bashrc, which, in some cases, is ran at the end of the bash_profile (so the old SCT will take precedent over the more recent SCT). What is the motivation for changing this behaviour?   A few years ago, SCT path was indeed set in the bash_profile, but for reason X(*), it was then set to bashrc. 

(*) i was hoping to find the issues/PR with the justification but after 5min search i gave up, sorry

Reply:
For sure. I mentioned this in my summary above.

My explanation is in the commit message for https://github.com/neuropoly/spinalcordtoolbox/pull/3125/commits/f80ff180af6b364a17b58eb084479e5c8c070213

Using `~/.bashrc`, someone running batch or remote jobs would have a hard time with SCT, because while `ssh user@server.com` will load ~/.bashrc, `ssh user@server.com sct_run_batch ... ` won't. And usually neither will jobs loaded via cron, or `at`. Our CI system *is a batch job*, but we didn't notice the problem because we had our CI script manually doing part (but not all) -- editing $PATH -- of what `./install_sct` does to configure SCT.

The problem became more obvious while working on this because

1. some of the platforms ended up not installing SCT in "in-place" mode, so .ci.sh's attempt to edit PATH failed
2. then, some platforms failed when I switched to loading from the RC file, because some refuse to load from ~/.bashrc when run non-interactively.

I decided to agree with the distros on this one: ~/.bashrc is only for interactive sessions.

interactive (or not) shells and login (or not) shells are a subtle unix thing. Something for the [unix hater's handbook](https://web.mit.edu/~simsong/www/ugh.pdf). But it'll be easier if we conform.

Reply:
I know this definitely definitely will cause some issues. Right now we just do

https://github.com/neuropoly/spinalcordtoolbox/blob/d64af23b69dbc396322cdc963a90a899b73736e0/install_sct#L677-L682

Which I think is pretty fragile anyway.

Maybe as a follow up to #3102 we can add something to `./install_sct` that checks if SCT is already loaded *before* installing and

a. warns the user and/or
b. greps every possible shellrc for the old PATH and tries to erase it

Reply:
Another fix is, since I forced the in-place install anyway (https://github.com/neuropoly/spinalcordtoolbox/pull/3125/commits/1b464c2450d2c9833375284e8365f1b13c5d6575 -- because our package is not laid out as best as it could, we should be nesting `src/spinalcordtoolbox` to avoid python's inconsistent package path rules), I could revert https://github.com/neuropoly/spinalcordtoolbox/pull/3125/commits/f80ff180af6b364a17b58eb084479e5c8c070213 on this branch, but put it aside for a later PR that we can write a good migration notice to our users about it.

Reply:
This is what is shipped on current Ubuntu:

```
$ cat /etc/skel/.profile 
# ~/.profile: executed by the command interpreter for login shells.
# This file is not read by bash(1), if ~/.bash_profile or ~/.bash_login
# exists.
# see /usr/share/doc/bash/examples/startup-files for examples.
# the files are located in the bash-doc package.

# the default umask is set in /etc/profile; for setting the umask
# for ssh logins, install and configure the libpam-umask package.
#umask 022

# if running bash
if [ -n "$BASH_VERSION" ]; then
    # include .bashrc if it exists
    if [ -f "$HOME/.bashrc" ]; then
	. "$HOME/.bashrc"
    fi
fi

# set PATH so it includes user's private bin if it exists
if [ -d "$HOME/bin" ] ; then
    PATH="$HOME/bin:$PATH"
fi

# set PATH so it includes user's private bin if it exists
if [ -d "$HOME/.local/bin" ] ; then
    PATH="$HOME/.local/bin:$PATH"
fi
```

```
$ cat /etc/skel/.bashrc 
# ~/.bashrc: executed by bash(1) for non-login shells.
# see /usr/share/doc/bash/examples/startup-files (in the package bash-doc)
# for examples

# If not running interactively, don't do anything
case $- in
    *i*) ;;
      *) return;;
esac

# don't put duplicate lines or lines starting with space in the history.
# See bash(1) for more options
HISTCONTROL=ignoreboth

# append to the history file, don't overwrite it
shopt -s histappend

# for setting history length see HISTSIZE and HISTFILESIZE in bash(1)
HISTSIZE=1000
HISTFILESIZE=2000

# check the window size after each command and, if necessary,
# update the values of LINES and COLUMNS.
shopt -s checkwinsize

# If set, the pattern "**" used in a pathname expansion context will
# match all files and zero or more directories and subdirectories.
#shopt -s globstar

# make less more friendly for non-text input files, see lesspipe(1)
[ -x /usr/bin/lesspipe ] && eval "$(SHELL=/bin/sh lesspipe)"

# set variable identifying the chroot you work in (used in the prompt below)
if [ -z "${debian_chroot:-}" ] && [ -r /etc/debian_chroot ]; then
    debian_chroot=$(cat /etc/debian_chroot)
fi

# set a fancy prompt (non-color, unless we know we "want" color)
case "$TERM" in
    xterm-color|*-256color) color_prompt=yes;;
esac

# uncomment for a colored prompt, if the terminal has the capability; turned
# off by default to not distract the user: the focus in a terminal window
# should be on the output of commands, not on the prompt
#force_color_prompt=yes

if [ -n "$force_color_prompt" ]; then
    if [ -x /usr/bin/tput ] && tput setaf 1 >&/dev/null; then
	# We have color support; assume it's compliant with Ecma-48
	# (ISO/IEC-6429). (Lack of such support is extremely rare, and such
	# a case would tend to support setf rather than setaf.)
	color_prompt=yes
    else
	color_prompt=
    fi
fi

if [ "$color_prompt" = yes ]; then
    PS1='${debian_chroot:+($debian_chroot)}\[\033[01;32m\]\u@\h\[\033[00m\]:\[\033[01;34m\]\w\[\033[00m\]\$ '
else
    PS1='${debian_chroot:+($debian_chroot)}\u@\h:\w\$ '
fi
unset color_prompt force_color_prompt

# If this is an xterm set the title to user@host:dir
case "$TERM" in
xterm*|rxvt*)
    PS1="\[\e]0;${debian_chroot:+($debian_chroot)}\u@\h: \w\a\]$PS1"
    ;;
*)
    ;;
esac

# enable color support of ls and also add handy aliases
if [ -x /usr/bin/dircolors ]; then
    test -r ~/.dircolors && eval "$(dircolors -b ~/.dircolors)" || eval "$(dircolors -b)"
    alias ls='ls --color=auto'
    #alias dir='dir --color=auto'
    #alias vdir='vdir --color=auto'

    alias grep='grep --color=auto'
    alias fgrep='fgrep --color=auto'
    alias egrep='egrep --color=auto'
fi

# colored GCC warnings and errors
#export GCC_COLORS='error=01;31:warning=01;35:note=01;36:caret=01;32:locus=01:quote=01'

# some more ls aliases
alias ll='ls -alF'
alias la='ls -A'
alias l='ls -CF'

# Add an "alert" alias for long running commands.  Use like so:
#   sleep 10; alert
alias alert='notify-send --urgency=low -i "$([ $? = 0 ] && echo terminal || echo error)" "$(history|tail -n1|sed -e '\''s/^\s*[0-9]\+\s*//;s/[;&|]\s*alert$//'\'')"'

# Alias definitions.
# You may want to put all your additions into a separate file like
# ~/.bash_aliases, instead of adding them here directly.
# See /usr/share/doc/bash-doc/examples in the bash-doc package.

if [ -f ~/.bash_aliases ]; then
    . ~/.bash_aliases
fi

# enable programmable completion features (you don't need to enable
# this, if it's already enabled in /etc/bash.bashrc and /etc/profile
# sources /etc/bash.bashrc).
if ! shopt -oq posix; then
  if [ -f /usr/share/bash-completion/bash_completion ]; then
    . /usr/share/bash-completion/bash_completion
  elif [ -f /etc/bash_completion ]; then
    . /etc/bash_completion
  fi
fi
```

The first thing their ~/.profile does is load ~/.bashrc, which means *installing to ~/.profile takes precedence*. It also means that using ~/.bashrc has always seemed like it's worked for us, even though if you read the bash and zsh manpages it's clear that they expect the two options weren't intended to overlap (see again: unix haters handbook).

Reply:
But one bug I did just notice for sure here: 

> This file is not read by bash(1), if ~/.bash_profile or ~/.bash_login exists.

So using ~/.bash_profile on Ubuntu will break our users systems in confusing ways. After they install SCT, but only after they've rebooted, they'll find that their tab completions and coloured CLIs are suddenly gone. That won't be very nice of us.

Uhhh. I guess we should be improving

https://github.com/neuropoly/spinalcordtoolbox/blob/d64af23b69dbc396322cdc963a90a899b73736e0/install_sct#L211-L230

so it detects which file the user is *actually* using.

---

In #1526 we should be able to make plain `pip install` work, and that will install to ~/.local/bin/, which most systems have in their PATH already -- Ubuntu does anyway; I think macOS does too? Then we can get out of the business of modifying people's shell configs without really telling them what's going on. And for people not on those systems we can just give a note about "if you installed SCT to your $HOME, make sure to add this PATH to your ~/.profile"

Reply:
I've convinced myself to pull this out for later: https://github.com/neuropoly/spinalcordtoolbox/issues/3149

Reply:
Reverted: 67271f6323db16e4cb0b2b9540ecca5b37b39abc

Reply:
Uhhhh weird. That made the installation break somehow?

Reply:
(3 days later) Ok I see what's going on. The old version didn't try to actually use `isct_*` like the programs they are, it just looked at them. Instead, we load the venv which gives us access to all the `scripts/`, i.e. `sct_*` and they get installed to `python/envs/venv_sct/bin` and that is added to `$PATH` by loading the venv.  But we *additionally* copy them to `$SCT_DIR/bin`, which is sort of misleading:


https://github.com/neuropoly/spinalcordtoolbox/blob/3dfd4f81f9e4364dab4d94894ea687a625dbdea5/install_sct#L631-L632

So how do the `isct_*` programs fit in? Well, instead of just editing `os.environ` to add `$SCT_DIR/bin` to PATH, we have this imitation version of the PATH algorithm here:

https://github.com/neuropoly/spinalcordtoolbox/blob/3dfd4f81f9e4364dab4d94894ea687a625dbdea5/spinalcordtoolbox/utils/sys.py#L274-L292

which is essentially only here for the 6 `isct_*` programs, and behaves differently than the usual PATH algorithm -- it chooses the last, not the first.


When `.ci.sh` tries my new, stronger, check (see also https://github.com/neuropoly/spinalcordtoolbox/pull/3125#discussion_r551057733) 

https://github.com/neuropoly/spinalcordtoolbox/blob/67271f6323db16e4cb0b2b9540ecca5b37b39abc/.ci.sh#L87

it fails because it `isct_*` aren't installed programs, from the conventional unix perspective. And the `sct_*` programs would fail too, because the venv hasn't been loaded yet.

---

The upshot of all this is:

I could

1. go back to using `ls` and say yolo
2. move the `conda activate venv_sct` line *before* this loop; then the `sct_*` scripts could be validated
3. move `conda activate venv_sct` before, but also retrofit a bunch of things to get rid of the hand-roll PATH algoriithm, and to make sure all the `isct_*` apps end up installed inside the python venv.
3. drop the whole check

I'm inclined to drop the whole thing. The unit tests should make installation glitches obvious enough -- they'll say "FileNotFoundError: isct_antsApplyTransforms" or something along those lines. Especially in light of the maintenance burden (https://github.com/neuropoly/spinalcordtoolbox/pull/3125#discussion_r551057733). 3 seems like the right long term solution, but that falls under #1526. It's actually the thing I've been procrastinating on for that the most, because it's hard.

Reply:
Dropped in https://github.com/neuropoly/spinalcordtoolbox/commit/7c8d606cfa36923bfb39543008da4eb23bdb0718